MSC4233: Remembering which server a user knocked through

turt2live opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/travis/msc/knocking-server/proposals/4233-remembering-knock-via.md)

Author: @turt2live, @ara4n 

----

I am Director of Standards Development at The Matrix.org Foundation C.I.C., Matrix Spec Core Team (SCT) member, employed by Element, and operate the t2bot.io service. This proposal is written and published with my role as a member of the SCT.

turt2live committed:
MSC: Remembering which server a user knock through

turt2live commented on proposals/4233-remembering-knock-via.md:
Implementation requirements:

* [ ] Client(?)
* [ ] Server

turt2live committed:
ed

ara4n commented on proposals/4233-remembering-knock-via.md:
```diff
@@ -0,0 +1,113 @@
+            {"type": "m.room.name", "state_key": "", "content": {"name": "My Room"}}
+          ]
+        },
+        "knock_servers": [ // new property
```
i wonder if `via` might be clearer?

ara4n commented on proposals/4233-remembering-knock-via.md:
```diff
@@ -0,0 +1,113 @@
+Servers may not have already tracked this, so information will be lacking for already-knocked rooms.
+Clients should expect errors, per usual, when attempting to call the summary API or any other endpoint.
+
+## Alternatives
```
presumably another option would be for the server to silently remember the knock server on behalf of the client and add an implicit via when they provide a raw room id?

turt2live replied:
The MSC already requires the server to track it, so I guess it's an open question for whether we want to continue using `via` for things, or have servers remember stuff.

turt2live commented on proposals/4233-remembering-knock-via.md:
This MSC also applies to rescinding knocks: servers [don't know](https://github.com/element-hq/synapse/blob/d80cd57c54427687afcb48740d99219c88a0fff1/synapse/handlers/room_member.py#L2022) what server to `make/send_leave` through remotely.

anoadragon453 replied:
This proposal currently doesn't help in the case of the user attempting to rescind a knock (see https://github.com/element-hq/synapse/issues/18030 for some discussion) as [`POST /_matrix/client/v3/rooms/{roomId}/leave`](https://spec.matrix.org/v1.13/client-server-api/#post_matrixclientv3roomsroomidleave) does not have a `via` parameter. It did not need one originally, as before knocking it was assumed the user had to be in the room already in order to `/leave`.

To support rescinding knocks over federation, either this proposal adds a `via` parameter to `/leave` as well, or we just make adding remotes entirely server-side as https://github.com/matrix-org/matrix-spec-proposals/pull/4233/files#r1862610665 describes.

turt2live replied:
`/leave` is also used to decline invites, which is also subject to the same sort of issues (if the sender's server is down). 

Overall, if we do want to add `via` to `/leave`, I'd ask that it be a dedicated MSC. If we want to move away from client-provided `via`s, that may also be a different MSC.

ara4n committed:
rewrite MSC4233 to require the server to remember the knock_server

...rather than putting the onus on the clients.
Done in consultation with Travis (I'm not just hijacking his MSC :P)
